diff --git a/src/ports/SkFontHost_win.cpp b/src/ports/SkFontHost_win.cpp
index 91283eb7135..ba9adf141d8 100644
--- a/src/ports/SkFontHost_win.cpp
+++ b/src/ports/SkFontHost_win.cpp
@@ -943,10 +943,13 @@ void SkScalerContext_GDI::generateFontMetrics(SkFontMetrics* metrics) {
         return;
     }
 
+// Alignment of considerable amounts of text is set too high with outline text metric values.
+// Standard text metrics for ascent and descent work better and provide results similar to other renderers.
+
 #ifndef SK_GDI_ALWAYS_USE_TEXTMETRICS_FOR_FONT_METRICS
     metrics->fTop = SkIntToScalar(-otm.otmrcFontBox.top);
-    metrics->fAscent = SkIntToScalar(-otm.otmAscent);
-    metrics->fDescent = SkIntToScalar(-otm.otmDescent);
+    metrics->fAscent = ((SkIntToScalar(-fTM.tmAscent) + SkIntToScalar(-otm.otmAscent)) / 2);
+    metrics->fDescent = ((SkIntToScalar(fTM.tmDescent) + SkIntToScalar(-otm.otmDescent)) / 2);
     metrics->fBottom = SkIntToScalar(-otm.otmrcFontBox.bottom);
     metrics->fLeading = SkIntToScalar(otm.otmLineGap);
     metrics->fAvgCharWidth = SkIntToScalar(otm.otmTextMetrics.tmAveCharWidth);
