diff --git a/BUILD.gn b/BUILD.gn
index 025203d66c7..bedd266923b 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -2367,9 +2367,9 @@ template("run_mksnapshot") {
         "--turbo-profiling-input",
         rebase_path(v8_builtins_profiling_log_file, root_build_dir),
 
-        # Replace this with --warn-about-builtin-profile-data to see the full
-        # list of builtins with incompatible profiles.
-        "--abort-on-bad-builtin-profile-data",
+        # Replace this with --abort-on-bad-builtin-profile-data to abort
+        # compilation when there are builtins with incompatible profiles.
+        "--warn-about-builtin-profile-data",
       ]
 
       if (!v8_enable_builtins_profiling && v8_enable_builtins_reordering) {
